src/python/pants/jvm/target_types.py

```diff
@@ -259,3 +260,41 @@ class DeployJarTarget(Target):
         "The JAR will contain class files for both first-party code and "
         "third-party dependencies, all in a common directory structure."
     )
+
+
+# -----------------------------------------------------------------------------------------------
+# `jvm_war` targets
+# -----------------------------------------------------------------------------------------------
+
+
+class JvmWarDependenciesField(Dependencies):
+    pass
+
+
+class JvmWarDescriptorAddressField(StringField):
+    alias = "descriptor"
+    help = "Address of a `file` target containing the descriptor (i.e., web.xml) for this WAR file."
```

Comment:
So does this need to be a single file? If so, it'd be helpful to add this:

> \n\nNote that this must be a single `file` target, not the target generator `files`. If you are using the `files` generator, use the address for the generated target, like `assets/logo.png` or `assets/logo.png:files`. See {doc_url('targets#target-generation')} for more on how the address is formed.

Reply:
Actually, why do this via a `file` target rather than `SingleSourceField`?

Reply:
> Actually, why do this via a `file` target rather than `SingleSourceField`?

No particular reason. I didn't even consider it. :(

Reply:
Cool, that'll be nice cleanup. 

I don't fault you for not considering it - we actually went very out of our way to "fix" `pex_binary` to not have a `source` field all the way back in Pants 2.2 iirc! It was conceptually broken, that the metadata for `pex_binary` describes the PEX to be built, not the file; that's what `python_source` is for. Generally, we don't want package types to have a `source` field.

The only reason it seems like it's okay here is that it's a `web.xml` file that solely exists for the sake of config for the `war_binary` iiuc.

Reply:
The reason to not to do it is we ever wanted `web.xml` to be supplied by some form of templating engine triggered by a plugin that could codegen to `FileSourceField`. But can't think of why would need that currently.

Reply:
Also see `docker_image`. Readme.com is down, but it has a `source` field and an `instructions` field where we generate it for you.